### PR TITLE
Add MinimalStdlib build-script product, build it in main CI jobs and PR testing jobs

### PR DESCRIPTION
**Pull request by kubamracek**
This adds the existing "minimal freestanding" stdlib variant preset (which builds a different, stripped-down version of the stdlib+runtime with minimal dependencies) into the main CI workflows, including PR testing that gates merging. The point of this is to block PRs that would regress the build of this "minimal freestanding" stdlib from merging.

For now, this doesn't run any tests, just builds. Testing will for now still only happen in the dedicated CI job (https://ci.swift.org/view/Dashboard/job/oss-swift-test-stdlib-with-toolchain-minimal/).

rdar://106456196

### CODE REVIEW

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

File: utils/swift_build_support/swift_build_support/products/minimalstdlib.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        self.cmake_options.define('SWIFT_PRIMARY_VARIANT_ARCH:STRING', 'x86_64')
+        self.cmake_options.define('SWIFT_PRIMARY_VARIANT_SDK:STRING', 'FREESTANDING')
+        self.cmake_options.define('SWIFT_HOST_TRIPLE:STRING', 'x86_64-apple-macosx10.13')
+        self.cmake_options.define('SWIFT_HOST_VARIANT', 'macosx')
+        self.cmake_options.define('SWIFT_HOST_VARIANT_ARCH', 'x86_64')
+        self.cmake_options.define('SWIFT_HOST_VARIANT_SDK', 'OSX')
+        self.cmake_options.define('SWIFT_STDLIB_ASSERTIONS:BOOL', 'FALSE')
+
+        # Configure build to only build the stdlib, and only a static version of it
+        self.cmake_options.define('SWIFT_BUILD_DYNAMIC_SDK_OVERLAY:BOOL', 'FALSE')
```
Comment (kubamracek) [inline utils/swift_build_support/swift_build_support/products/minimalstdlib.py:73]: Note to self: This needs to somehow be unified with the settings in build-preset.ini

**Comment by kubamracek** [comment]
@swift-ci please clean test

**Review by tbkka — APPROVED** [review]

**Review by mikeash — APPROVED** [review]

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
From the PR testing build bot:

```
Build Percentage 	 Build Duration (sec) 	 Build Phase
================ 	 ==================== 	 ===========
27.1%             	 4212.93               	 macosx-x86_64-swift-test
20.1%             	 3128.13               	 macosx-x86_64-swift-build
6.0%              	 935.97                	 Running tests for swiftpm
4.9%              	 753.91                	 Running tests for swiftdriver
4.1%              	 638.08                	 Building swiftpm
4.1%              	 635.57                	 Running tests for swiftdocc
3.9%              	 602.32                	 macosx-x86_64-lldb-test
2.7%              	 424.7                 	 Building sourcekitlsp
2.6%              	 408.26                	 Running tests for swiftsyntax
2.3%              	 352.62                	 Building swiftsyntax
2.3%              	 350.8                 	 Building skstresstester
2.2%              	 341.94                	 Building llvm
2.2%              	 334.92                	 Building swiftformat
2.1%              	 323.74                	 Running tests for swiftformat
2.1%              	 320.19                	 Installing swiftpm
1.9%              	 301.6                 	 Building swiftevolve
1.8%              	 278.34                	 Building swiftdocc
1.8%              	 273.26                	 Installing swiftdocc
1.1%              	 171.65                	 Building swiftdriver
1.0%              	 159.79                	 Building minimalstdlib    // <<<=================
0.7%              	 110.14                	 Building earlyswiftdriver
0.6%              	 91.58                 	 Running tests for sourcekitlsp
0.5%              	 80.03                 	 Building earlyswiftsyntax
0.5%              	 78.48                 	 macosx-x86_64-lldb-build
0.4%              	 58.79                 	 macosx-x86_64-llbuild-build
0.3%              	 48.44                 	 Building indexstoredb
0.2%              	 29.27                 	 Running tests for indexstoredb
0.1%              	 18.85                 	 macosx-x86_64-swift-install
0.1%              	 18.11                 	 Building playgroundsupport
0.1%              	 13.88                 	 Running tests for skstresstester
0.1%              	 13.01                 	 macosx-x86_64-llbuild-test
```

So it takes an extra 2 minutes 40 seconds to build. CC @shahmishal @DougGregor @tbkka -- that feels like totally worth the value it adds?

**Review by ktoso — APPROVED** [review]
Thanks this looks great

I'd _really_ want this to be added, it is very tricky to get all these #ifs right while working on concurrency types, the additional validation will save a lot of time and fire-fighting from mistakes

**Comment by al45tair** [comment]
Agreed, 2 minutes 40 seconds is well worth it.

**Review by shahmishal — APPROVED** [review]

**Review by al45tair — APPROVED** [review]

**Review by etcwilde — COMMENTED** [review]
From a code perspective, this looks fine. Isn't mostly a matter of being a different build configuration than an entirely separate product though? I suppose the static builds of Foundation and dispatch are kind of the same way, so maybe it's fine.

**Comment by kubamracek** [comment]
I'm going to merge this as is, because the value seems very high right now, and follow up separately on the remaining things.

**Comment by shahmishal** [comment]
@kubamracek  Can you cherry-pick this on to release/5.9?

**Comment by kubamracek** [comment]
@shahmishal sure! https://github.com/apple/swift/pull/64677

**Pull request merged**